### ISSUE TITLE
[MPS] multinomial(p, 1) intermittently returns zero-probability indices — exponential_() emits exact -0.0, Gumbel fast path turns it into a NaN that argmax selects

### ISSUE THREAD

**Issue by JohnMachado11** [pytorch/pytorch#192577]
### 🐛 Describe the bug

**Expected:** `torch.multinomial(p, 1)` never returns an index `i` with `p[i] == 0`.

**Actual:** on MPS it does so intermittently — ~1 per ~220 draws for a 152k-category distribution whose tail underflows to exact zeros (i.e., any temperature-sharpened softmax, any top-k/top-p filtered distribution, any masked policy). CPU: zero occurrences.

**Repro (~60 s, no model):**

```python
import torch
V = 151936
logits = torch.randn(V) * 2.0
logits[12345] += 200.0                    # tails underflow to exact 0.0
p = torch.softmax(logits, -1).to("mps")
bad = sum(float(p[int(torch.multinomial(p, 1))]) == 0.0 for _ in range(20_000))
print(bad)   # M1 Pro, macOS 15.7.7, torch 2.13.0: 90.  CPU: 0.
```

**Root cause** (each step measured; full forensic dataset available — ~477k individually verified draws, 695 corruption events, saved tensors):

1. MPS `exponential_(1)` emits exact zeros at ~3.3e-8/element (census: 10 zeros in 3.04e8 samples; CPU: 0 in the same count). Captured values are `-0.0` (signbit set), the fingerprint of a negated log at a boundary uniform. `Distributions.cpp` already documents this hazard class — *"on CUDA side there is a protection to avoid 0s"* — MPS has no such protection.
2. The n=1 Gumbel fast path (`q = exponential_; q = p/q; argmax`, `aten/src/ATen/native/Distributions.cpp` ~594–628) computes `0 / -0 = NaN` when the RNG zero lands on a zero-probability entry (one element per affected draw; over a 152k vocab that's ~0.5% of draws).
3. MPS argmax propagates NaN as the maximum, so the returned "sample" is the position of the RNG zero — an index unrelated to the distribution. (When the zero lands on a positive-probability entry, `p / -0 = -inf` loses argmax and the defect is invisible.)
4. Fully deterministic given RNG state: restoring the state reproduces the identical corrupt index (45/45 replays); `torch.mps.synchronize()` has no effect; reproduces at rest. Not a race.

**Impact:** silent wrong-token emission in any sampled decoding on Apple Silicon (`do_sample` / top-k / top-p; greedy and beam are unaffected) at ~1 per few hundred sampled tokens. Discovered via an LLM agent that began investigating objects that didn't exist — the model's own distribution wanted the correct token at 99.9% while the sampler emitted a token with recorded probability exactly 0.0; traces available.

**Suggested fix:** protect the exponential output *before* the division — e.g. clamp the MPS `exponential_` kernel's output to a positive minimum, mirroring the CUDA protection. (Post-division guards don't work: `clamp` propagates the NaN, and `nan_to_num` would also mask genuine `p` NaNs that the current `_assert_async` checks are designed to catch.)

**Workaround for users:** draw on CPU — `torch.multinomial(p.cpu(), 1)`.

Relates to #136623 (same op, different failure surface: that issue is out-of-domain indices from the multi-sample kernel; this one is in-domain zero-probability indices from the single-sample fast path).

### Versions

<details>

```
Collecting environment information...
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.7.7 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.4 (v3.12.4:8e8a4baf65, Jun  6 2024, 17:33:18) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-15.7.7-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
```

</details>

cc @pbelevich @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @albanD

**Comment by Srinandha-Murugesan** [comment]
Hello, I was able to reproduce this and I have changes ready based on clamping the exponential_ kernel output similar to CUDA along with tests for checking if the situation's fixed along with a specific RNG seed I found that had the same issue. Could this be marked as actionable so that I can make a PR? Thank you!

**Comment by JohnMachado11** [comment]
> Hello, I was able to reproduce this and I have changes ready based on clamping the exponential_ kernel output similar to CUDA along with tests for checking if the situation's fixed along with a specific RNG seed I found that had the same issue. Could this be marked as actionable so that I can make a PR? Thank you!


Great to hear it's been reproduced. As far as making a PR, please do, it's actionable. Not sure what else you would need from me. 

Just link to this issue when the PR is up.

**Comment by Srinandha-Murugesan** [comment]
I believe the maintainers have to attach the actionable label to the issue before I can open the PR. 

I made a draft one for now: https://github.com/pytorch/pytorch/pull/192621

**Comment by JohnMachado11** [comment]
> I believe the maintainers have to attach the actionable label to the issue before I can open the PR. 
> 
> I made a draft one for now: https://github.com/pytorch/pytorch/pull/192621


Ah, okay. Makes sense.

Thanks for the PR. Hopefully a maintainer sees this soon. 👍

**Comment by Isalia20** [comment]
I think this is actionable. It has real world use case that you outlined (LLM inference) so PR for it makes sense. cc: @malfet

**Comment by malfet** [comment]
Not for triage review: CPU exponential_() might be suffering from the same issue... At least on ARM

**Comment by janeyx99** [comment]
This looks to be a subset of the problem in https://github.com/pytorch/pytorch/issues/192812 already, closing in favor of tracking there. Comment if that is not true!

**Comment by Srinandha-Murugesan** [comment]
@janeyx99 just wanted to clarify, this issue is for MPS and I have an open PR being reviewed by @Isalia20 and @malfet to fix this here: https://github.com/pytorch/pytorch/pull/192621

#192812 is a separate CPU-specific issue that was found in the PR. It's unrelated to the MPS changes that close this so I think the issue should be kept open. If not, should I link my PR to https://github.com/pytorch/pytorch/issues/192812?

**Comment by JohnMachado11** [comment]
> This looks to be a subset of the problem in [#192812](https://github.com/pytorch/pytorch/issues/192812) already, closing in favor of tracking there. Comment if that is not true!

I agree with @Srinandha-Murugesan  that this should stay open. 

[#192812](https://github.com/pytorch/pytorch/issues/192812) is very closely related (both invoke `torch.multinomial`)  but #192812 is a `cpu` and `float16` specific problem. 

Whereas _this_ issue is `mps` related only.

**Comment by janeyx99** [comment]
Ah, I was about to reopen but the commit fixes this so I’ll keep it closed. Thanks!

**Comment by JohnMachado11** [comment]
> Ah, I was about to reopen but the commit fixes this so I’ll keep it closed. Thanks!

Yep, all good.

Happy to have contributed to the legendary `PyTorch`. :)
